File: java/ReverseWordsInAString.java
public class ReverseWordsInAString {
    public String reverseWords(String s) {
        if (s == null) {
            return null;
        }
        String rev = new StringBuilder(s).reverse().toString();
        StringBuilder res = new StringBuilder("");
        String token = new String("");
        int n = rev.length();
        for (int i = 0; i < n; ++i) {
            while (i < n && rev.charAt(i) == ' ') {
                ++i;
            }
            if (i == n) {
                break;
            }
            int j = i;
            while (j < n && rev.charAt(j) != ' ') {
                ++j;
            }
            int k = j;
            --j;
            StringBuilder word = new StringBuilder(rev.substring(i, k)).reverse();
            res.append(token + word);
            token = " ";
            i = k - 1;
        }
        return res.toString();
    }

    public static void main(String[] args) {
        ReverseWordsInAString solver = new ReverseWordsInAString();
        String s = null;
        s = new String("the sky is blue");
        System.out.println(solver.reverseWords(s));
    }
}

File: python/reverse_words_in_a_string.py
#!/usr/bin/env python3

class Solution:
    # @param s, a string
    # @return a string
    def reverseWords(self, s):
        if s is None:
            return None
        s = s.strip()
        n = len(s)
        ls = list(s[::-1])
        i = 0
        while i < n:
            while i < n and ls[i] == ' ':
                i += 1
            if i == n:
                break
            j = i
            while j < n and ls[j] != ' ':
                j += 1
            ls[i:j] = ls[i:j][::-1]
            i = j
        return ''.join(ls)

def main():
    solver = Solution()
    print(solver.reverseWords("the sky is blue"))

if __name__ == '__main__':
    main()
